packages/shylu/shylu_dd/bddc/src/shylu_PComm.h

```diff
@@ -0,0 +1,486 @@
+#ifndef PCOMMBDDC_H
+#define PCOMMBDDC_H
+
+#include <stdio.h>
+#include <iostream>
+#include <fstream>
+#include <string.h>
+#include <sstream>
+#include <iomanip>
+#include <stdlib.h>
+#include <algorithm>
+#include <vector>
+#include <cmath>
+#include <mpi.h>
+
+#include "Tpetra_Distributor.hpp"
+#include "shylu_errorBDDC.h"
+
+using Teuchos::RCP;
+using Teuchos::rcp;
+
+namespace bddc {
+
+template <class LO, class GO, class SX> class PComm
+{
+public:
+  
+  PComm(RCP< PComm<LO,GO,SX> > & pComm,
+	RCP<const Teuchos::Comm<int> > Comm)
+  {
+    m_distributor = rcp( new Tpetra::Distributor(Comm) );
+    m_distributor->createFromSendsAndRecvs(pComm->m_distributor->getProcsTo(),
+					   pComm->m_distributor->getProcsFrom());
+    m_sourceLength = pComm->m_sourceLength;
+    m_targetLength = pComm->m_targetLength;
+    m_sendCount = pComm->m_sendCount;
```

Comment:
I'm not keen on this "break encapsulation by pulling all the fields out of the input" design pattern.

Reply:
Did you have in mind using getter functions to not break encapsulation? I could do that once we have a chance to discuss.

Reply:
@crdohrm Let's talk more about this in the coming week.

Reply:
Sounds good. I should be in this week Monday through Thursday.

Reply:
I think I understand.  This is a copy constructor that copies everything except the `MPI_Comm` and the Distributor.  Why not just write a copy constructor?  You don't need to take the input pComm by RCP because you don't keep it.

Reply:
This code just deep copies the input pComm.  Take the input pComm by const reference instead of by RCP.

Reply:
We discussed this over the phone.